Strategy (#146)
rust-unofficial/patterns: patterns/strategy.md

# Strategy (aka Policy)

## Description

The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concerns in object-oriented software design.
It also allows to achieve decoupling software modules known as [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
The basic idea behind Strategy is given an algorithm, solving a particular problem, we define only skeleton of the algorithm at abstract level, but delegate specific algorithm’s implementation or some of its parts to another class or method.

> A client using algorithm may choose a specific implementation, while the general algorithm workflow remains the same.

Reviewer: We need spacing in between paragraph, one large chuck of text is hard to read.

In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.
This is why we call it Dependency Inversion.

## Motivation

Imagine we are working on a project that generates reports every month. 

> We need the reports to be generated in different formats, e.g., in `JSON` or `Plain Text` formats. 

Reviewer: ```suggestion
We need the reports to be generated in different formats (strategies), e.g., in `JSON` or `Plain Text` formats. 
```

We should relate it back to "strategy".

But thing very and we don't know what kind of requiremnt we may get in the future. 
For example, we may need to generate our report in a completly new format, 
or just modify the existing format. 
So, in order to make chages to our existing code base as much as possible 
while keeping our solution adoptable to such changes, 
we might exploit Strategy design pattern. 

> Rust has powerful trait system which allows to implement this pattern. 

Reviewer: Is this line necessary?

## Example

In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`. 
`Text` and `Json` are our strategy structs. 
These strategies have to implement `Formatter` trait.

```rust
use std::fmt::Write as FmtWrite;
use std::{error, result};
type Result = result::Result<String, Box<dyn error::Error>>;

> struct Context {

Reviewer: I think we should not use `Context` as an example. Too many places uses the word "context" especially in golang which may make people confuse what is "context". We should have a more specific word for this situation.

    pub keys: Vec<String>,
    pub values: Vec<i32>,
}

trait Formatter {
    fn run(&self, context: &Context) -> Result;
}

struct Report;

impl Report {
    fn generate<T: Formatter>(g: T) -> Result {

>         //perform here backend operations which should not bother caller...
>         //fetch data from database

Reviewer: ```suggestion
        // backend operations...
```

        let keys = vec!["one".to_string(), "two".to_string()];

>         let values = vec![1, 2];

Reviewer: Shouldn't we show a case with `HashMap` for this? If we use two vecs, it means that it could have different length.

Author: Shouldn't length of keys and length of values should be the equal? We assume each value corresponds to each key. I think we it is not that important whether we use `HashMap` or another collection. The main focus should be how we apply Strategy. After all, it is a toy example. I am also planning mention `serde` with use example.

Author: After the second thought, using `HashMap` allows getting rid of `Context` and ugly `for` loop on vecs.

        // generate report
        g.run(&Context { keys, values })
    }
}

struct Text;
impl Formatter for Text {
    fn run(&self, context: &Context) -> Result {

>         let mut s = String::new();

Reviewer: This part should take `&mut String` rather than creating string every time it runs.

Author: Do you mean `run(&self, context: &Context, s: &mut String) -> Result` ?

Reviewer 2: why? If that's only for performance reason..come on, it's just an example! :)

            if s.len() > 1 {
                write!(&mut s, ",")?;
            }
            write!(&mut s, "{{\"{}\":\"{}\"}}", key, val)?;
        }
        write!(&mut s, "]")?;
        Ok(s)
    }
}

fn main() {
    assert_eq!(
        String::from("one 1\ntwo 2\n"),
        Report::generate(Text).unwrap()
    );

    assert_eq!(
        String::from(r#"[{"one":"1"},{"two":"2"}]"#),
        Report::generate(Json).unwrap()
    );
}

```

## Advantages

Separation of concerns. In the previous example, Report does not know anything about specific implementations of `Json` and `Text`, whereas the output implementations does not care about how data is preprocessed, stored, and fetched. 
The only thing they have to know is context and and a specific trait and method to implement, i.e., `Formatter` and `run`. 

## Disadvantages

For each strategy there must me implemented at least one module, so number of modules increases with number of strategies.
If there are many strategies to choose from then users have to know how strategies differ from one another.

## Discussion

In the previous example all strategies in a single file.

> Typically in Rust, each strategy should be implemented in a separate module file.

Reviewer: ```suggestion
Ways of providing different strategies includes:

- All in one file (as shown in this example)
- Separated as modules, E.g. `formatter::json` module, `formatter::text` module
- Use compiler feature flags, E.g. `json` feature, `text` feature 
- Separated as crates, E.g. `json` crate, `text` crate
```

We may need to link each individual method, especially the use compiler feature flags.

# Strategy (aka Policy)

## Description

The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concerns in object-oriented software design.

> It also allows to achieve decoupling software modules known as [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle).

Reviewer 2: ```suggestion
It also allows to decouple software modules through [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
```
Do you prefer it this way?

> The basic idea behind Strategy is given an algorithm, solving a particular problem, we define only skeleton of the algorithm at abstract level, but delegate specific algorithm’s implementation or some of its parts to another class or method.

Reviewer 2: ```suggestion
The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem, we define only the skeleton of the algorithm at an abstract level and we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.
```

> A client using algorithm may choose a specific implementation, while the general algorithm workflow remains the same.

Reviewer 2: ```suggestion
In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
```

In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.
This is why we call it Dependency Inversion.

## Motivation

Imagine we are working on a project that generates reports every month. 
We need the reports to be generated in different formats (strategies), e.g., in `JSON` or `Plain Text` formats. 
But things very and we don't know what kind of requirement we may get in the future. 
For example, we may need to generate our report in a completly new format, 

> or just modify the existing format. 

Reviewer 2: ```suggestion
or just modify the existing format.

```

# Strategy (aka Policy)

## Description

> The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concerns in object-oriented software design.

Reviewer 2: ```suggestion
The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern)
is a technique that enables separation of concerns in object-oriented software design.
```

It also allows to decouple software modules through [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).

> The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem, we define only the skeleton of the algorithm at an abstract level and we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.

Reviewer 2: ```suggestion
The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem,
we define only the skeleton of the algorithm at an abstract level and
we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.
```

In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.

> In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.

Reviewer 2: ```suggestion
In other words, the abstract specification of the class does not depend on the specific implementation of the derived class,
but specific implementation must adhere to the abstract specification.
```

This is why we call it "Dependency Inversion".

## Motivation

Imagine we are working on a project that generates reports every month.

> We need the reports to be generated in different formats (strategies), e.g., in `JSON` or `Plain Text` formats. 

Reviewer 2: ```suggestion
We need the reports to be generated in different formats (strategies), e.g. `JSON` or `Plain Text`. 
```

> But things very and we don't know what kind of requirement we may get in the future.

Reviewer 2: ```suggestion
But things vary over time and we don't know what kind of requirement we may get in the future.
```

For example, we may need to generate our report in a completly new format,

> or just modify the existing format.

Reviewer 2: ```suggestion
or just modify one of the existing formats.
```

## Example

> In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`
> `Text` and `Json` are our strategy structs.

Reviewer 2: ```suggestion
In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
while `Text` and `Json` are our strategy structs.
```

> These strategies have to implement `Formatter` trait.

Reviewer 2: ```suggestion
These strategies have to implement the `Formatter` trait.
```

This is why we call it "Dependency Inversion".

## Motivation

Imagine we are working on a project that generates reports every month.
We need the reports to be generated in different formats (strategies), e.g.,
in `JSON` or `Plain Text` formats.
But things vary over time and we don't know what kind of requirement we may get in the future.
For example, we may need to generate our report in a completly new format,
or just modify one of the existing formats.

## Example

In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
while `Text` and `Json` are our strategy structs.
These strategies have to implement the `Formatter` trait.

```rust
use std::collections::HashMap;
type Data = HashMap<String, u32>;

trait Formatter {
    fn run(&self, data: &Data) -> String;
}

struct Report;

impl Report {
    fn generate<T: Formatter>(g: T) -> String {
        // backend operations...
        let mut data = HashMap::new();
        data.insert("one".to_string(), 1);
        data.insert("two".to_string(), 2);
        // generate report
        g.run(&data)
    }
}

struct Text;
impl Formatter for Text {
    fn run(&self, data: &Data) -> String {

>         data
>             .iter()

Reviewer 2: ```suggestion
        data.iter()
```
This is how rustfmt likes it

In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.
This is why we call it "Dependency Inversion".

## Motivation

Imagine we are working on a project that generates reports every month.
We need the reports to be generated in different formats (strategies), e.g.,
in `JSON` or `Plain Text` formats.
But things vary over time and we don't know what kind of requirement we may get in the future.
For example, we may need to generate our report in a completly new format,
or just modify one of the existing formats.

## Example

In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
while `Text` and `Json` are our strategy structs.
These strategies have to implement the `Formatter` trait.

```rust
use std::collections::HashMap;

> use std::fmt::Write as FmtWrite;

Reviewer 2: ```suggestion
use std::fmt::Write;
```

This has no impact, why do you do it?

Author: Probably leftover. :-)

trait Formatter {
    fn run(&self, data: &Data) -> String;
}

struct Report;

impl Report {
    fn generate<T: Formatter>(g: T) -> String {
        // backend operations...
        let mut data = HashMap::new();
        data.insert("one".to_string(), 1);
        data.insert("two".to_string(), 2);
        // generate report
        g.run(&data)
    }
}

struct Text;
impl Formatter for Text {
    fn run(&self, data: &Data) -> String {

>         let s: String = data
>             .iter()
>             .map(|(key, val)| format!("{} {}\n", key, val))
>             .collect();
>         return s;

Reviewer 2: ```suggestion
        data
            .iter()
            .map(|(key, val)| format!("{} {}\n", key, val))
            .collect()
```

>             write!(&mut s, "{} {}\n", key, val)?;

Reviewer 2: ```suggestion
            let entry = format!("{} {}\n", key, val);
            s.push_str(&entry);
```
You use `write!` that may fail, but if you use something like this you are sure that it doesn't fail.
So basically all the `Result`s can be removed.

Reviewer 2: This `Text`, in particular, could be written really elegantly with `map` I believe, but that's not required, it's just a thought I had.

> use std::error;
> type Data = HashMap<String, u32>;
> 
> trait Formatter {
>     fn run(&self, data: &Data) -> Result<String, Box<dyn error::Error>>;
> }
> 
> struct Report;
> 
> impl Report {
>     fn generate<T: Formatter>(g: T) -> Result<String, Box<dyn error::Error>> {
>         // backend operations...
>         let mut data = HashMap::new();
>         data.insert("one".to_string(), 1);
>         data.insert("two".to_string(), 2);
>         // generate report
>         g.run(&data)
>     }
> }
> 
> struct Text;
> impl Formatter for Text {
>     fn run(&self, data: &Data) -> Result<String, Box<dyn error::Error>> {
>         let mut s = String::new();
> 
>         for (key, val) in data {
>             write!(&mut s, "{} {}\n", key, val)?;
>         }
>         Ok(s)
>     }
> }
> 
> struct Json;
> impl Formatter for Json {
>     fn run(&self, data: &Data) -> Result<String, Box<dyn error::Error>> {

Reviewer 2: ```suggestion
use std::error::Error;
type Data = HashMap<String, u32>;

trait Formatter {
    fn run(&self, data: &Data) -> Result<String, Box<dyn Error>>;
}

struct Report;

impl Report {
    fn generate<T: Formatter>(g: T) -> Result<String, Box<dyn Error>> {
        // backend operations...
        let mut data = HashMap::new();
        data.insert("one".to_string(), 1);
        data.insert("two".to_string(), 2);
        // generate report
        g.run(&data)
    }
}

struct Text;
impl Formatter for Text {
    fn run(&self, data: &Data) -> Result<String, Box<dyn Error>> {
        let mut s = String::new();

        for (key, val) in data {
            write!(&mut s, "{} {}\n", key, val)?;
        }
        Ok(s)
    }
}

struct Json;
impl Formatter for Json {
    fn run(&self, data: &Data) -> Result<String, Box<dyn Error>> {
```

In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.
This is why we call it "Dependency Inversion".

## Motivation

Imagine we are working on a project that generates reports every month.
We need the reports to be generated in different formats (strategies), e.g.,
in `JSON` or `Plain Text` formats.
But things vary over time and we don't know what kind of requirement we may get in the future.
For example, we may need to generate our report in a completly new format,
or just modify one of the existing formats.

## Example

In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
while `Text` and `Json` are our strategy structs.
These strategies have to implement the `Formatter` trait.

```rust
use std::collections::HashMap;
use std::fmt::Write as FmtWrite;
use std::{error, result};

> type Result = result::Result<String, Box<dyn error::Error>>;

Reviewer 2: I think that this is confusing. I would prefer to use `Result<String, Box<dyn error::Error>>` everywhere it is necessary, instead of creating a type alias.
What other reviewers think?

Reviewer 2: Maybe we can call it something like `AnyResult` and we can remove the `use std::{error, result};` line?

Author: I added `use std::{error, result};` to avoid compiler errors because I used `Result`. Now that your offer `AnyResult` we can remove `use std::{error, result};`.

Author: Other reviewers hasn't sent yet anything concerning the alias for `Result<String, Box<dyn error::Error>>`. If it is really confusing I can remove and use `Result<String, Box<dyn error::Error>>`.

Reviewer 2: yes, sorry, only `Result` is in the prelude apparently

    fn run(&self, data: &Data) -> Result;
}

struct Report;

impl Report {
    fn generate<T: Formatter>(g: T) -> Result {
        // backend operations...
        let mut data = HashMap::new();
        data.insert("one".to_string(), 1);
        data.insert("two".to_string(), 2);
        // generate report
        g.run(&data)
    }
}

struct Text;
impl Formatter for Text {
    fn run(&self, data: &Data) -> Result {
        let mut s = String::new();

        for (key, val) in data {
            write!(&mut s, "{} {}\n", key, val)?;
        }
        Ok(s)
    }
}

struct Json;
impl Formatter for Json {
    fn run(&self, data: &Data) -> Result {
        let mut s = String::from("[");

        for (key, val) in data {
            if s.len() > 1 {
                write!(&mut s, ",")?;
            }
            write!(&mut s, "{{\"{}\":\"{}\"}}", key, val)?;
        }
        write!(&mut s, "]")?;
        Ok(s)
    }
}

fn main() {
    let s = Report::generate(Text).unwrap();
    assert!(s.contains("one 1"));
    assert!(s.contains("two 2"));

    let s = Report::generate(Json).unwrap();
    assert!(s.contains(r#"{"one":"1"}"#));
    assert!(s.contains(r#"{"two":"2"}"#));
}

> 

Reviewer 2: ```suggestion
```

## Advantages

> Separation of concerns. In the previous example, Report does not know anything about specific

Reviewer 2: ```suggestion
Separation of concerns. In the previous example, `Report` does not know anything about specific
```

> Separation of concerns. In the previous example, `Report` does not know anything about specific

Reviewer 2: ```suggestion
The main advantage is separation of concerns. For example, in this case `Report` does not know anything about specific
```

implementations of `Json` and `Text`, whereas the output implementations does not care about how data is
preprocessed, stored, and fetched.

> The only thing they have to know is context and and a specific trait and method to implement,

Reviewer 2: ```suggestion
The only thing they have to know is context and a specific trait and method to implement,
```

i.e,`Formatter` and `run`.

## Disadvantages

> For each strategy there must me implemented at least one module, so number of modules

Reviewer 2: ```suggestion
For each strategy there must be implemented at least one module, so number of modules
```

increases with number of strategies.
If there are many strategies to choose from then users have to know how strategies differ
from one another.

## Discussion

> In the previous example all strategies in a single file.

Reviewer 2: ```suggestion
In the previous example all strategies are implemented in a single file.
```

- All in one file (as shown in this example, similar to being separated as modules)
- Separated as modules, E.g. `formatter::json` module, `formatter::text` module
- Use compiler feature flags, E.g. `json` feature, `text` feature
- Separated as crates, E.g. `json` crate, `text` crate

However, we don't need to use traits in order to design this pattern in Rust.

The following toy example demonstrates the idea of the Strategy pattern using Rust
`closures`:

```rust
struct Adder;
impl Adder {
    pub fn add<F>(x: u8, y: u8, f: F) -> u8
    where
        F: Fn(u8, u8) -> u8,
    {
        f(x, y)
    }
}

fn main() {
    let arith_adder = |x, y| x + y;
    let bool_adder = |x, y| {
        if x == 1 || y == 1 {
            1
        } else {
            0
        }
    };
    let custom_adder = |x, y| 2 * x + y;

    assert_eq!(9, Adder::add(4, 5, arith_adder));
    assert_eq!(0, Adder::add(0, 0, bool_adder));
    assert_eq!(5, Adder::add(1, 3, custom_adder));
}

> 

Reviewer 2: ```suggestion
```

```

In fact, Rust already uses this idea for `Options`'s `map` method

```rust
fn main() {
    let val = Some("Rust");
    let len_strategy = |s: &str| s.len();
    let first_byte_strategy = |s: &str| s.bytes().next().unwrap();

    assert_eq!(4, val.map(len_strategy).unwrap());
    assert_eq!(82, val.map(first_byte_strategy).unwrap());
}

> 

Reviewer 2: ```suggestion
```

>     let val = Some("Rust");
>     let len_strategy = |s: &str| s.len();
>     let first_byte_strategy = |s: &str| s.bytes().next().unwrap();
> 
>     assert_eq!(4, val.map(len_strategy).unwrap());
>     assert_eq!(82, val.map(first_byte_strategy).unwrap());

Reviewer 2: ```suggestion
    let val = Some("Rust");

    let len_strategy = |s: &str| s.len();
    assert_eq!(4, val.map(len_strategy).unwrap());

    let first_byte_strategy = |s: &str| s.bytes().next().unwrap();
    assert_eq!(82, val.map(first_byte_strategy).unwrap());
```

Reviewer 2: do you find it clearer like this?

Reviewer 2: (I just moved the instructions)

Author: 👍

- Use compiler feature flags, E.g. `json` feature, `text` feature
- Separated as crates, E.g. `json` crate, `text` crate

However, we don't need to use traits in order to design this pattern in Rust.

The following toy example demonstrates the idea of the Strategy pattern using Rust
`closures`:

```rust
struct Adder;
impl Adder {
    pub fn add<F>(x: u8, y: u8, f: F) -> u8
    where
        F: Fn(u8, u8) -> u8,
    {
        f(x, y)
    }
}

fn main() {
    let arith_adder = |x, y| x + y;
    let bool_adder = |x, y| {
        if x == 1 || y == 1 {
            1
        } else {
            0
        }
    };
    let custom_adder = |x, y| 2 * x + y;

    assert_eq!(9, Adder::add(4, 5, arith_adder));
    assert_eq!(0, Adder::add(0, 0, bool_adder));
    assert_eq!(5, Adder::add(1, 3, custom_adder));
}

```

In fact, Rust already uses this idea for `Options`'s `map` method

```rust
fn main() {
    let val = Some("Rust");

    let len_strategy = |s: &str| s.len();
    assert_eq!(4, val.map(len_strategy).unwrap());

    let first_byte_strategy = |s: &str| s.bytes().next().unwrap();
    assert_eq!(82, val.map(first_byte_strategy).unwrap());
}
```

## See also

> [Strategy Pattern](https://en.wikipedia.org/wiki/Strategy_pattern)
> 
> [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)

Reviewer 2: ```suggestion
- [Strategy Pattern](https://en.wikipedia.org/wiki/Strategy_pattern)
- [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)<!-- markdown-link-check-disable -->
- [Policy Based Design](https://en.wikipedia.org/wiki/Modern_C%2B%2B_Design#Policy-based_design)<!-- markdown-link-check-enable -->
```

Reviewer 2: let's try if this works..

# Strategy (aka Policy)

## Description

The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern)

> is a technique that enables separation of concerns in object-oriented software design.

Reviewer: Is this really related to object-oriented software design? I can't see how it relates.

Author: @pickfire We may discuss here on what OOD is for days and stil can't come to a consensus, can we?. Rust book ([chapter 17]( https://doc.rust-lang.org/book/ch17-01-what-is-oo.html)) writes:

> There is no consensus in the programming community about what features a language must have to be considered object oriented.
Nevertheless, in my opinion, if you are dealing with DATA + BEHAVIOUR then we traditionally call them "OBJECTS", and if we think in terms of "OBJECTS" then we simply call it object oriented.  

As for design patterns, they are simply a set of principles (or whatever you call them) to follow to make software design clear, easy to refactor, readable, etc.
The main point of this section is the Strategy pattern and not whether or not it relates to OOD or what OOD is.

Reviewer: Okay, I guess I am fine with it after you explained, but I rarely see the docs mentioning about objects which make me a bit confused.

>  The main point of this section is the Strategy pattern and not whether or not it relates to OOD or what OOD is.

Yes, that is why I mention this. Because could be done in non OOD as well, since it could be separated from file level.

It also allows to decouple software modules through [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).

> The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem,

Reviewer: ```suggestion

The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem,
```

A new line would be good to separate the basic idea.

we define only the skeleton of the algorithm at an abstract level and

> we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.

Reviewer: ```suggestion
we separate the specific algorithm’s implementation into different parts.
```

I think using "another class or method" here may confuse people, in rust "class" is not a (strong) concept but it is borrowed from other languages so not everyone can understand it.

In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
In other words, the abstract specification of the class does not depend on the specific implementation of the derived class,
but specific implementation must adhere to the abstract specification.
This is why we call it "Dependency Inversion".

## Motivation

Imagine we are working on a project that generates reports every month.
We need the reports to be generated in different formats (strategies), e.g.,
in `JSON` or `Plain Text` formats.
But things vary over time and we don't know what kind of requirement we may get in the future.
For example, we may need to generate our report in a completly new format,
or just modify one of the existing formats.

## Example

In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
while `Text` and `Json` are our strategy structs.
These strategies have to implement the `Formatter` trait.

```rust
use std::collections::HashMap;
type Data = HashMap<String, u32>;

trait Formatter {
    fn run(&self, data: &Data) -> String;
}

struct Report;

impl Report {

>     fn generate<T: Formatter>(g: T) -> String {

Reviewer: Ah, like I mentioned before. It's just an example. Since we are in rust patterns, I think we should show a good pattern since others will be learning from here. I still think taking in `&mut String` should be used here, at least others not knowing this will be more aware about it, it won't make it much complicated right?

Reviewer 3: @fade2black Seems like this is the last review comment that we could address before merging, any thoughts on it?

Author: @simonsan  I am on it.

Author: @simonsan @pickfire Fixed.

Reviewer 2: I think that taking the string as mut is not always the best approach.
The rust library is full of functions that returns strings as output.
Taking it as mut gives you flexibility, but it is more error prone and less readable. It depends on your priorities.
Anyway, I don't think it matters that much, so it's fine like this :)

Author: Whether it is `&mut String` or `String` as return, I don't care very much in this example. This is just an example.
Depending on use cases both `&mut String` and `String` as return would work fine. No need to waste time on [trivial things](https://en.wikipedia.org/wiki/Law_of_triviality).

        // backend operations...
        let mut data = HashMap::new();
        data.insert("one".to_string(), 1);
        data.insert("two".to_string(), 2);
        // generate report
        g.run(&data)
    }
}

struct Text;
impl Formatter for Text {

>     fn run(&self, data: &Data) -> String {

Reviewer: Just my two cents, it would be better to change `run` to `format`, since it is named formatter, not really a pattern but run is very general.

        data.iter()
            .map(|(key, val)| format!("{} {}\n", key, val))
            .collect()
    }
}

struct Json;
impl Formatter for Json {
    fn run(&self, data: &Data) -> String {

>         let mut s = String::from("[");
>         let mut entry;
> 
>         for (key, val) in data {
>             if s.len() > 1 {
>                 s.push(',');
>             }
>             entry = format!("{{\"{}\":\"{}\"}}", key, val);
>             s.push_str(&entry);
>         }
>         s.push(']');
>         return s;

Reviewer: ```suggestion
        let mut s = String::from("[");
        let mut iter = data.into_iter();
                            
        if let Some((key, val)) = data.next() {
            let entry = format!(r#"{"{}":"{}"}"#, key, val);
            s.push_str(&entry);
            while let Some((key, val)) = data.next() {
                s.push(',');   
                let entry = format!(r#"{"{}":"{}"}"#, key, val);
                s.push_str(&entry);
            }
        }
        s.push(']');
        s
```
This should be more efficient since we unrolled the loop to remove the branch, ideally we want to make it branchless. Like the above example, it may be a bit longer but it is more efficient.

There is also another method to this, we could collect into `Vec<String>` and `slices::join` them, although it requires double allocation which could make it shorter but less efficient maybe.

Author: I had in mind but didn't know how to do it in Rust. Good idiom: create iter, first next, and then iter til the end.

Author: not `data.next()`, but `iter.next()`

    let s = Report::generate(Json);
    assert!(s.contains(r#"{"one":"1"}"#));
    assert!(s.contains(r#"{"two":"2"}"#));
}
```

## Advantages

The main advantage is separation of concerns. For example, in this case `Report` does not know anything about specific
implementations of `Json` and `Text`, whereas the output implementations does not care about how data is
preprocessed, stored, and fetched.
The only thing they have to know is context and a specific trait and method to implement,
i.e,`Formatter` and `run`.

## Disadvantages

For each strategy there must be implemented at least one module, so number of modules
increases with number of strategies.
If there are many strategies to choose from then users have to know how strategies differ
from one another.

## Discussion

In the previous example all strategies are implemented in a single file.
Ways of providing different strategies includes:

- All in one file (as shown in this example, similar to being separated as modules)
- Separated as modules, E.g. `formatter::json` module, `formatter::text` module
- Use compiler feature flags, E.g. `json` feature, `text` feature
- Separated as crates, E.g. `json` crate, `text` crate

> Serde crate is a good example of the `Strategy` pattern in action. Serde allows [full customization](https://serde.rs/custom-serialization.html)
> of the serialization behavior by manually implementing `Serialize` and `Deserialize` traits for our type.

Reviewer: Would be good to have an example for serde strategy.

```suggestion
Serde crate is a good example of the `Strategy` pattern in action. Serde allows [full customization](https://serde.rs/custom-serialization.html)
of the serialization behavior by manually implementing `Serialize` and `Deserialize` traits for our type.
For example, we could easily swap `serde_json` with `serde_cbor` since they expose a similar method.
Having this makes the helper crate `serde_transcode` much more useful and ergonomic.
```

implementations of `Json` and `Text`, whereas the output implementations does not care about how data is
preprocessed, stored, and fetched.
The only thing they have to know is context and a specific trait and method to implement,
i.e,`Formatter` and `run`.

## Disadvantages

For each strategy there must be implemented at least one module, so number of modules
increases with number of strategies.
If there are many strategies to choose from then users have to know how strategies differ
from one another.

## Discussion

In the previous example all strategies are implemented in a single file.
Ways of providing different strategies includes:

- All in one file (as shown in this example, similar to being separated as modules)
- Separated as modules, E.g. `formatter::json` module, `formatter::text` module
- Use compiler feature flags, E.g. `json` feature, `text` feature
- Separated as crates, E.g. `json` crate, `text` crate

Serde crate is a good example of the `Strategy` pattern in action. Serde allows [full customization](https://serde.rs/custom-serialization.html)
of the serialization behavior by manually implementing `Serialize` and `Deserialize` traits for our type.

> For example, we could easily swap `serde_json` with `serde_cbor` since they expose a similar method.

Reviewer: Sorry, grammatical error I put just now.
```suggestion
For example, we could easily swap `serde_json` with `serde_cbor` since they expose similar methods.
```

Having this makes the helper crate `serde_transcode` much more useful and ergonomic.

However, we don't need to use traits in order to design this pattern in Rust.

The following toy example demonstrates the idea of the Strategy pattern using Rust
`closures`:

```rust
struct Adder;
impl Adder {
    pub fn add<F>(x: u8, y: u8, f: F) -> u8
    where
        F: Fn(u8, u8) -> u8,
    {
        f(x, y)
    }
}

fn main() {
    let arith_adder = |x, y| x + y;
    let bool_adder = |x, y| {
        if x == 1 || y == 1 {
            1
        } else {
            0
        }
    };
    let custom_adder = |x, y| 2 * x + y;

    assert_eq!(9, Adder::add(4, 5, arith_adder));
    assert_eq!(0, Adder::add(0, 0, bool_adder));
    assert_eq!(5, Adder::add(1, 3, custom_adder));
}

```

> In fact, Rust already uses this idea for `Options`'s `map` method

Reviewer: ```suggestion
In fact, Rust already uses this idea for `Options`'s `map` method:
```

```rust
fn main() {
    let val = Some("Rust");

    let len_strategy = |s: &str| s.len();
    assert_eq!(4, val.map(len_strategy).unwrap());

    let first_byte_strategy = |s: &str| s.bytes().next().unwrap();
    assert_eq!(82, val.map(first_byte_strategy).unwrap());
}
```

## See also

- [Strategy Pattern](https://en.wikipedia.org/wiki/Strategy_pattern)
- [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)<!-- markdown-link-check-disable -->

> - [Policy Based Design](https://en.wikipedia.org/wiki/Modern_C%2B%2B_Design#Policy-based_design)<!-- markdown-link-check-enable -->

Reviewer: Weird, why do we need to disable this link check?

Author: @MarcoIeni ?

> - [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)<!-- markdown-link-check-disable -->
> - [Policy Based Design](https://en.wikipedia.org/wiki/Modern_C%2B%2B_Design#Policy-based_design)<!-- markdown-link-check-enable -->

Reviewer 3: ```suggestion
- [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)
- [Policy Based Design](https://en.wikipedia.org/wiki/Modern_C++_Design#Policy-based_design)
```

Reviewer 3: Can we try that and see if it works, otherwise we revert and keep the disabled link check

Reviewer 3: (Would be an upstream issue if it works like this)

# Strategy (aka Policy)

## Description

> The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concern in object-oriented software design. It also allows to achieve decoupling software modules known as [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle). The basic idea behind Strategy is given an algorithm, solving a particular problem, we define only skeleton of the algorithm at abstract level, but delegate specific algorithm’s implementation or some of its parts to another class or method. A client using algorithm may choose a specific implementation, while the general algorithm workflow remains the same. In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification. This is why we call it Dependency Inversion.

Reviewer 2: ```suggestion
The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concerns in object-oriented software design. It also allows to achieve decoupling software modules known as [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle). The basic idea behind Strategy is given an algorithm, solving a particular problem, we define only skeleton of the algorithm at abstract level, but delegate specific algorithm’s implementation or some of its parts to another class or method. A client using algorithm may choose a specific implementation, while the general algorithm workflow remains the same. In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification. This is why we call it Dependency Inversion.
```

Reviewer 2: It's difficult to do review and visualize diffs on single lines. Can you go to new line on every `.` or something like that?

Author: Do you mean typing 

 _This is sentence 1. 
 This is sentence 2.
 This is sentence 3._

Instead of 

_This is sentence 1. This is sentence 2. This is sentence 3._ ?

Reviewer 2: Yes

Author: Of course, if it does not violate formatting rules.

## Motivation

> Suppose we are working on a project whose purpose is periodically read big data, [wrangle](https://en.wikipedia.org/wiki/Data_wrangling) it, and finally store it in a database as key-value pairs. After the data are stored in database, we need to provide conumers with APIs to query data. However, different consumers may want data in different formats (json, txt, xml, etc), even in new one we have never seen so far. But we don't know in advance in what format consumers would query data. So, every time a request for a new format implementation comes, we have to rebuild our software. We'd also like to change our code as little as possible. Ideally, the wrangling part of our softaware shouldn't change because it has nothing to with data serialization. This is a one scenario where we could apply the Strategy pattern to design our software. We could aslo publish the wrangling part as a library leaving the serialization implementation to library users. The following example explains the idea.

Reviewer: This example is quite huge, maybe we should use a simpler one but show `serde-json` as a use case on how the formatting is done using either pretty formatter or compact formatter for a bigger example?

Author: I agree.

Author: BTW, I don't use `gofmt`. I use `rustfmt 1.4.18-stable`. As for indentation, I use VS Code, and I set number of spaces to 2. `rust fmt` fixes it to four.

> Suppose we are working on a project whose purpose is periodically read big data, [wrangle](https://en.wikipedia.org/wiki/Data_wrangling) it, and finally store it in a database as key-value pairs. After the data are stored in database, we need to provide conumers with APIs to query data. However, different consumers may want data in different formats (json, txt, xml, etc), even in new one we have never seen so far. But we don't know in advance in what format consumers would query data. So, every time a request for a new format implementation comes, we have to rebuild our software. We'd also like to change our code as little as possible. Ideally, the wrangling part of our softaware shouldn't change because it has nothing to with data serialization. This is a one scenario where we could apply the Strategy pattern to design our software. We could aslo publish the wrangling part as a library leaving the serialization implementation to library users. The following example explains the idea.

Reviewer 3: ```suggestion
Suppose we are working on a project whose purpose is periodically read big data, [wrangle](https://en.wikipedia.org/wiki/Data_wrangling) it, and finally store it in a database as key-value pairs.
After the data are stored in database, we need to provide conumers with APIs to query data.
However, different consumers may want data in different formats (json, txt, xml, etc), even in new one we have never seen so far. But we don't know in advance in what format consumers would query data.
So, every time a request for a new format implementation comes, we have to rebuild our software.
We'd also like to change our code as little as possible.
Ideally, the wrangling part of our softaware shouldn't change because it has nothing to with data serialization.
This is a one scenario where we could apply the Strategy pattern to design our software.
We could aslo publish the wrangling part as a library leaving the serialization implementation to library users. The following example explains the idea.
```

## Example

```rust
// This trait must be implemented by different formats
pub trait Formatter {
  fn run(&self, report: &Report);
}

// STRATEGIES:

> // Implementation for Json format

Reviewer: This is an example that have no focus on performance, we should state that the implementation below is inefficient and should not be used.

Author: Do you mean dynamic dispatch against static?

Reviewer: No, I meant the way it uses `print!`.

Author: @pickfire I have removed `prints`. Added `assert_eq!`s.

mod vendor1 {

>   use super::*;

Reviewer: Can this be formatted to 4 spaces?

Author: Yes, of course.

// Instead, it provides high level abstraction
// - vector of Strings: `keys`
// - vector of integers: `values`
// - method: `generate`
// 
// In other words, Report does not depend on format implementation,
// but format implementations depend on Report (abtraction). This is called 
// Dependency Inversion Principle. 

pub struct Report {
  pub keys:   Vec<String>,
  pub values: Vec<i32>,
  // User must provide an object which implements Generator trait
  formatter: Box<dyn Formatter>
}

impl Report {
  pub fn format(&self) {
    self.formatter.run(&self);
  }

  pub fn new(formatter: Box<dyn Formatter>) -> Self {
    let (keys, values) = Self::data_from_db();
    Report{ keys, values, formatter }
  }

  fn data_from_db() -> (Vec<String>, Vec<i32>) {
    let keys = vec!["one".to_string(), "two".to_string()];
    let values =  vec![1, 2];
    (keys, values)
  }
}

fn main() {
  let json_report = Report::new(Box::new(vendor1::Json));
  let text_report = Report::new(Box::new(vendor2::Text));

  println!("JSON format:");
  json_report.format();
  
  println!("\n");

  println!("Plain text format:");
  text_report.format();
}

```

## Advantages

> Separation of concerns. In the previous example, Report does not know anything about specific implementations of `Vendor1::Json` and `Vendor2::Text`, whereas the output implementations does not care about how data is preprocessed, stored, and fetched. The only thing they have to know is context and and a specific trait and method to implement, i.e., `Generator` and `run`. 

Reviewer 3: ```suggestion
Separation of concerns. In the previous example, Report does not know anything about specific implementations of `Vendor1::Json` and `Vendor2::Text`, whereas the output implementations does not care about how data is preprocessed, stored, and fetched. 
The only thing they have to know is context and and a specific trait and method to implement, i.e., `Generator` and `run`. 
```

## Disadvantages

> For each strategy there must me implemented at least one module, so number of modules increases with number of strategies. If there are many strategies to choose from then users have to know how strategies differ from one another.

Reviewer 3: ```suggestion
For each strategy there must me implemented at least one module, so number of modules increases with number of strategies.
If there are many strategies to choose from then users have to know how strategies differ from one another.
```

## Discussion

> In the previous example all strategies in a single file. Typically in Rust, each strategy should be implemented in a separate module file. However, we don't need to use traits in order to design this pattern in Rust. The following toy example demonstates the idea of the Strategy pattern using Rust closures

Reviewer 3: ```suggestion
In the previous example all strategies in a single file.
Typically in Rust, each strategy should be implemented in a separate module file.
However, we don't need to use traits in order to design this pattern in Rust.

The following toy example demonstrates the idea of the Strategy pattern using Rust `closures`:
```

```rust
struct Adder;
impl Adder{
  pub fn add<F>(x: u8, y: u8, f: F) -> u8
      where F: Fn(u8, u8) -> u8 {
      f(x,y)
  }
}

fn main() {
  let arith_adder = |x, y| x + y;

>   let bool_adder = |x, y| if x==1 || y==1 {1} else {0};

Reviewer: I believe this needs to be formatted.

Author: Do you mean spaces? Or one liner `if x==1...`

Reviewer: Both spaces and that.

  let custom_adder = |x, y| 2*x + y;

  println!("{:?}", Adder::add(4, 5, arith_adder));
  println!("{:?}", Adder::add(0, 0, bool_adder));
  println!("{:?}", Adder::add(0, 3, custom_adder));
}
```

In fact, Rust already uses this idea for `Options`'s `map` method

```rust
fn main() {
  let val = Some("Rust");
  let len_strategy = |s: &str| s.len();
  let first_byte_strategy = |s: &str| s.bytes().next().unwrap();

  println!("len: {}", val.map(len_strategy).unwrap());
  println!("first bite: {}", val.map(first_byte_strategy).unwrap());
}

> ```

Reviewer: Maybe we could mention `serde-json` as an example? Like one could just swap out `serde-json` with `toml` easily.

Author: Good idead.
